scenarios/kubernetes_e2e.py

```diff
@@ -44,49 +45,169 @@ def check(*cmd):
+                        continue
+                    self.env_files.append(self.parse_env(line))
+
+    def add_gce_ssh(self, priv, pub):
+        """Copies priv, pub keys to $WORKSPACE/.ssh."""
+        gce_ssh = '%s/.ssh/google_compute_engine' % self.workspace
+        gce_pub = '%s/.ssh/google_compute_engine.pub' % self.workspace
+        shutil.copy(priv, gce_ssh)
+        shutil.copy(pub, gce_pub)
+        self.add_environment(
+            'JENKINS_GCE_SSH_PRIVATE_KEY_FILE=%s' % gce_ssh,
+            'JENKINS_GCE_SSH_PUBLIC_KEY_FILE=%s' % gce_pub,
+        )
+
+    def add_service_account(self, path):
+        """Sets GOOGLE_APPLICATION_CREDENTIALS to path."""
+        self.add_environment('GOOGLE_APPLICATION_CREDENTIALS=%s' % path)
+
+    @property
+    def runner(self):
+        """Finds the best version of e2e-runner.sh."""
+        options = ['e2e-runner.sh', test_infra('jenkins/e2e-image/e2e-runner.sh')]
+        for path in options:
+            if os.path.isfile(path):
+                return path
+        raise ValueError('Cannot find e2e-runner at any of %s' % ', '.join(options))
+
+
+    def install_prerequisites(self):
+        """Copies upload-to-gcs and kubetest if needed."""
+        parent = os.path.dirname(self.runner)
+        if not os.path.isfile(os.path.join(parent, 'upload-to-gcs.sh')):
+            shutil.copy(
+                test_infra('../kubernetes/hack/jenkins/upload-to-gcs.sh'),
+                os.path.join(parent, 'upload-to-gcs.sh'))
+        if not os.path.isfile(os.path.join(parent, 'kubetest')):
+            check('go', 'install', 'k8s.io/test-infra/kubetest')
+            shutil.copy(
+                os.path.expandvars('${GOPATH}/bin/kubetest'),
+                os.path.join(parent, 'kubetest'))
+
+    def start(self):
+        """Runs e2e-runner.sh after setting env and installing prereqs."""
+        env = {}
+        env.update(self.env_files)
+        env.update(self.env)
+        self.install_prerequisites()
+        check_env(env, self.runner)
+
+
+class DockerMode(object):
+    """Runs e2e tests via docker run kubekins-e2e."""
+    def __init__(self, container, workspace, sudo, tag, mount_paths):
+        self.tag = tag
+        try:  # Pull a newer version if one exists
+            check('docker', 'pull', kubekins(tag))
+        except subprocess.CalledProcessError:
+            pass
+
+        print 'Starting %s...' % container
+
+        self.container = container
+        self.cmd = [
+            'docker', 'run', '--rm',
+            '--name=%s' % container,
+            '-v', '%s/_artifacts:/workspace/_artifacts' % workspace,
+            '-v', '/etc/localtime:/etc/localtime:ro',
+        ]
+        for path in mount_paths:
+            self.cmd.extend(['-v', path])
+
+        if sudo:
+            self.cmd.extend(['-v', '/var/run/docker.sock:/var/run/docker.sock'])
+        self.add_environment(
+            'HOME=/workspace',
+            'WORKSPACE=/workspace')
```

Comment:
Yeah, I suspect this file will require additional futzing to work on prow.

In this particular instance the image should already set WORKSPACE=/workspace, so it shouldn't matter.